pkg/istio-agent/agent.go

```diff
@@ -298,6 +314,62 @@ func (a *Agent) initializeEnvoyAgent(ctx context.Context) error {
 	return nil
 }
 
+func (a *Agent) generateGRPCBootstrap() error {
+	// generate metadata
+	node, err := a.generateNodeMetadata()
+	if err != nil {
+		return fmt.Errorf("failed generating node metadata: %v", err)
+	}
+	xdsMeta, err := structpb.NewStruct(node.RawMetadata)
+	if err != nil {
+		return fmt.Errorf("failed converting to xds metadata: %v", err)
+	}
+
+	// TODO secure control plane channel (most likely JWT + TLS, but possibly allow mTLS)
+	serverURI := a.proxyConfig.DiscoveryAddress
+	if a.cfg.ProxyXDSViaAgent {
+		serverURI = "localhost:15010"
+	}
+
+	// TODO use struct from gRPC lib if they ever exist
+	data := map[string]interface{}{
+		// TODO grab this port from somewhere else
+		"xds_servers": map[string]interface{}{
+			"server_uri": serverURI,
+			// TODO channel_creds
+		},
+		"node": corev3.Node{
+			Id:       node.ID,
+			Locality: node.Locality,
+			Metadata: xdsMeta,
+		},
+	}
+
+	if a.secOpts.OutputKeyCertToDir != "" {
+		data["certificate_providers"] = map[string]interface{}{
+			"default": map[string]interface{}{
+				"plugin_name": "file_watcher",
+				"config": pemfile.Options{
+					KeyFile:  path.Join(a.secOpts.OutputKeyCertToDir, "key.pem"),
+					CertFile: path.Join(a.secOpts.OutputKeyCertToDir, "cert-chain.pem"),
+					RootFile: path.Join(a.secOpts.OutputKeyCertToDir, "root-cert.pem"),
+					// TODO use a more appropriate interval
+					RefreshDuration: 15 * time.Minute,
+				},
+			},
+		}
+	}
+	jsonData, err := json.Marshal(data)
```

Comment:
does this work properly? since corev3.node is a proto, json.marshal may have unexpected results

Reply:
I see the right `json` tags in the Go struct. We could add tags to our own maybe? I didn't want to use protomarshal because the whole `data` would need to be compatible.